In [1]:
import pandas as pd

clean_df = pd.read_csv("../data/processed/clean_mandi_prices.csv")
clean_df['Price Date'] = pd.to_datetime(clean_df['Price Date'], format='%Y-%m-%d')

In [2]:
clean_df_raw = pd.read_csv("../data/processed/clean_mandi_prices.csv")
print(clean_df_raw['Price Date'].head())

0    2023-06-06
1    2023-06-06
2    2023-06-06
3    2023-06-06
4    2023-06-06
Name: Price Date, dtype: str


In [3]:
clean_df = clean_df_raw.copy()
clean_df['Price Date'] = pd.to_datetime(clean_df['Price Date'], format='%Y-%m-%d')

In [4]:
print(clean_df.dtypes['Price Date'])
print(clean_df['Price Date'].min(), clean_df['Price Date'].max())

datetime64[us]
2023-06-06 00:00:00 2025-06-11 00:00:00


In [5]:
clean_df['Year'] = clean_df['Price Date'].dt.year
clean_df['Month'] = clean_df['Price Date'].dt.month

In [6]:
clean_df[['Price Date', 'Year', 'Month']].head()

,Price Date,Year,Month
0,2023-06-06,2023,6
1,2023-06-06,2023,6
2,2023-06-06,2023,6
3,2023-06-06,2023,6
4,2023-06-06,2023,6


In [7]:
clean_df.groupby('Commodity')['Modal_Price'].mean()

Commodity
Onion     2214.632856
Potato    1357.409433
Rice      3229.150143
Tomato    3724.152976
Wheat     2353.474050
Name: Modal_Price, dtype: float64

In [8]:
clean_df.groupby('District Name')['Modal_Price'].mean().sort_values(ascending=False)

District Name
shamli           4201.428571
kanpur           4130.000000
gorakhpur        2241.266509
balrampur        2219.436920
lucknow          2151.811010
ghazipur         2123.467296
bahraich         2115.943878
mahoba           2104.645858
gonda            2099.491551
basti            2086.348898
varanasi         2079.132901
kaushambi        2077.268535
chandauli        2074.996622
deoria           2074.079945
ballia           2072.822836
unnao            2068.313084
kushinagar       2063.866071
barabanki        2058.956640
jaunpur          2057.046689
ambedkarnagar    2051.801802
pratapgarh       2041.469786
hardoi           2040.527492
azamgarh         2038.260734
banda            2018.694030
amethi           2011.938994
lalitpur         2008.900754
shravasti        2003.178634
jhansi           1975.817343
mathura          1975.306563
rampur           1965.278969
meerut           1964.346905
etawah           1959.129515
prayagraj        1957.864073
maharajganj      1953.614309


In [11]:
clean_df.groupby('Market Name')['Modal_Price'].mean().sort_values(ascending=False).head(10)

Market Name
Shamli            4201.428571
Uttaripura        4130.000000
Jhansi (Grain)    3470.000000
Richha            2864.264706
Panchpedwa        2590.630491
Banthara          2587.461988
Varanasi          2471.952297
Sehjanwa          2419.983498
Kishunpur         2323.544601
Badda             2314.285714
Name: Modal_Price, dtype: float64

In [12]:
clean_df.groupby('Market Name')['Modal_Price'].mean().sort_values(ascending=False).tail(10)

Market Name
Shikohabad       1235.077778
Hasanpur         1233.274380
Bhehjoi          1177.034247
Jalalabad        1173.571429
Dhanaura         1171.777883
Visoli           1165.685374
Dibiapur         1137.280405
Yusufpur         1113.636364
Parikshitgarh    1068.750000
Tanda(Rampur)     785.294118
Name: Modal_Price, dtype: float64

In [13]:
clean_df[clean_df['Market Name'] == 'Shamli']['Commodity'].value_counts()

Commodity
Onion    7
Name: count, dtype: int64

In [14]:
clean_df.groupby(['Market Name', 'Commodity'])['Modal_Price'].mean()

Market Name  Commodity
Achalda      Onion        2331.220183
             Potato       1355.322581
             Rice         3037.307692
             Tomato       2994.831933
             Wheat        2375.849515
                             ...     
Wazirganj    Rice         3335.555556
             Tomato       4553.281250
             Wheat        2377.555556
Yusufpur     Onion        2100.000000
             Potato       1015.000000
Name: Modal_Price, Length: 728, dtype: float64

In [15]:
clean_df.groupby(['Market Name', 'Commodity']).size().sort_values().head(10)

Market Name      Commodity
Yusufpur         Onion        1
Rampurmaniharan  Wheat        1
Jasra            Rice         1
Sehjanwa         Wheat        1
Mawana           Rice         1
Jhansi (Grain)   Rice         1
Fatehpur Sikri   Rice         1
Charkhari        Onion        1
Kishunpur        Onion        1
                 Potato       1
dtype: int64

Market-level groupings often have very few records (many combinations have just 1 row) — market-level price estimates are unreliable for sparse combinations. Crop-level and district-level aggregations are more statistically trustworthy. Dashboard should account for this (e.g., minimum sample size warning or fallback to district average)

In [16]:
monthly_avg = clean_df.groupby(['Commodity', 'Month'])['Modal_Price'].mean()

In [17]:
monthly_avg['Tomato']

Month
6     1658.580504
7     6541.917097
8     5168.324176
9     1781.274074
10    1242.065789
11    1535.594406
Name: Modal_Price, dtype: float64

In [18]:
print(monthly_avg['Tomato'])

Month
6     1658.580504
7     6541.917097
8     5168.324176
9     1781.274074
10    1242.065789
11    1535.594406
Name: Modal_Price, dtype: float64


In [19]:
print(len(monthly_avg['Tomato']))

6


In [20]:
monthly_avg.unstack(level='Commodity')

Commodity,Onion,Potato,Rice,Tomato,Wheat
Month,,,,,
1,1865.163121,1033.816815,NaN,NaN,2488.810329
2,1863.779889,887.839188,NaN,NaN,2508.665939
3,2046.215414,1099.006145,NaN,NaN,NaN
4,1598.450074,1268.695169,3225.695092,NaN,NaN
5,1396.686224,1347.982991,3229.722611,NaN,NaN
6,1191.915860,1345.542879,3239.610451,1658.580504,2250.693622
7,2050.750396,1637.819857,NaN,6541.917097,2257.563103
8,2327.885474,1669.955209,NaN,5168.324176,2283.536882
9,2728.038135,1605.265055,NaN,1781.274074,2312.439166


In [21]:
clean_df.groupby('Commodity')['Month'].nunique()

Commodity
Onion     12
Potato    12
Rice       3
Tomato     6
Wheat      9
Name: Month, dtype: int64

**Data limitation identified:** Month coverage varies significantly by crop:
- Onion, Potato: full 12-month coverage (reliable for seasonality analysis)
- Wheat: 9-month coverage
- Tomato: 6-month coverage (June-November only)
- Rice: 3-month coverage (April-June only)

This is very likely a data collection artifact (the source dataset may have been scraped/assembled with inconsistent commodity tracking windows), not a reflection of real-world crop availability. Seasonality analysis and price prediction for Tomato and Rice should be scoped/validated only within their available months. This will be revisited in Day 6 (feature engineering) and Day 8+ (model validation).

In [22]:
year_month_avg = clean_df.groupby(['Commodity', 'Year', 'Month'])['Modal_Price'].mean()

In [23]:
year_month_avg['Rice']

Year  Month
2025  4        3225.695092
      5        3229.722611
      6        3239.610451
Name: Modal_Price, dtype: float64

In [24]:
year_month_avg['Tomato']

Year  Month
2023  6        1658.580504
      7        6541.917097
      8        5168.324176
      9        1781.274074
      10       1242.065789
      11       1535.594406
Name: Modal_Price, dtype: float64

**Data limitation — updated finding:** Investigated WHY Tomato and Rice have partial month coverage by checking Year+Month combinations directly:
- Rice: ALL data is from 2025 only (April, May, June) — zero records from 2023 or 2024
- Tomato: ALL data is from 2023 only (June through November) — zero records from 2024 or 2025
- Onion, Potato: confirmed full 12-month coverage (needs verification across all 3 years, not just month count)
- Wheat: 9-month coverage (needs similar year-by-year verification)

Conclusion: this is a dataset construction/collection artifact — different crops were evidently scraped/tracked during different, non-overlapping time windows — NOT genuine crop seasonality. Rice and Tomato's monthly price figures reflect a single year's snapshot each, not a multi-year seasonal average. This means: (1) true year-over-year seasonality analysis is only valid for crops with multi-year coverage, (2) Rice/Tomato price prediction models (Day 8+) cannot be validated across multiple years and should be scoped/documented accordingly, (3) this may influence which crops are prioritized for the core ML demonstration in this project.

In [25]:
clean_df[clean_df['Commodity'] == 'Onion'].groupby('Year')['Month'].nunique()

Year
2023     7
2024    12
2025     5
Name: Month, dtype: int64

In [26]:
clean_df[(clean_df['Commodity'] == 'Onion') & (clean_df['Year'] == 2025)]['Month'].unique()

array([2, 3, 4, 5, 6], dtype=int32)

**Boundary check confirmed:** Onion's 7/12/5 month-per-year pattern (2023/2024/2025) is consistent with the dataset's overall date range (2023-06-06 to 2025-06-11) — 2023 only has June-Dec available (7 months) and 2025 only has Feb-June available (5 months, with January 2025 specifically missing — a minor, isolated gap, not investigated further). 2024 has full 12-month coverage. This is fundamentally different from Rice/Tomato's pattern, where entire YEARS are missing despite the dataset covering that period for other crops — confirming Rice/Tomato have a crop-specific collection artifact, while Onion's gaps are explained by the dataset's natural start/end boundary.

In [28]:
price_pivot = clean_df.pivot_table(
    values='Modal_Price',
    index='Commodity',
    columns='Month',
    aggfunc='mean'
)
price_pivot

Month,1,2,3,4,5,6,7,8,9,10,11,12
Commodity,,,,,,,,,,,,
Onion,1865.163121,1863.779889,2046.215414,1598.450074,1396.686224,1191.915860,2050.750396,2327.885474,2728.038135,3009.980661,3696.238591,2869.677125
Potato,1033.816815,887.839188,1099.006145,1268.695169,1347.982991,1345.542879,1637.819857,1669.955209,1605.265055,1543.595506,1513.269792,1369.281944
Rice,NaN,NaN,NaN,3225.695092,3229.722611,3239.610451,NaN,NaN,NaN,NaN,NaN,NaN
Tomato,NaN,NaN,NaN,NaN,NaN,1658.580504,6541.917097,5168.324176,1781.274074,1242.065789,1535.594406,NaN
Wheat,2488.810329,2508.665939,NaN,NaN,NaN,2250.693622,2257.563103,2283.536882,2312.439166,2344.151639,2440.834610,2461.444059


In [29]:
count_pivot = clean_df.pivot_table(
    values='Modal_Price',
    index='Commodity',
    columns='Month',
    aggfunc='count'
)
count_pivot

Month,1,2,3,4,5,6,7,8,9,10,11,12
Commodity,,,,,,,,,,,,
Onion,3102.0,6106.0,6838.0,6770.0,7346.0,4754.0,6306.0,6671.0,6477.0,6412.0,5960.0,3729.0
Potato,6982.0,7145.0,7323.0,7286.0,7702.0,7836.0,7272.0,7055.0,6908.0,6853.0,6442.0,6934.0
Rice,NaN,NaN,NaN,1630.0,2145.0,421.0,NaN,NaN,NaN,NaN,NaN,NaN
Tomato,NaN,NaN,NaN,NaN,NaN,2739.0,2907.0,546.0,675.0,684.0,143.0,NaN
Wheat,2246.0,458.0,NaN,NaN,NaN,2399.0,2694.0,2630.0,2589.0,2440.0,2219.0,2306.0


In [2]:
import pandas as pd
import numpy as np

clean_df = pd.read_csv("../data/processed/clean_mandi_prices.csv")
clean_df['Price Date'] = pd.to_datetime(clean_df['Price Date'], format='%Y-%m-%d')
clean_df['Year'] = clean_df['Price Date'].dt.year
clean_df['Month'] = clean_df['Price Date'].dt.month

print(clean_df.shape)
print(clean_df.dtypes['Price Date'])

(188080, 12)
datetime64[us]


In [3]:
volatility = clean_df.groupby('Commodity')['Modal_Price'].agg(['mean', 'std'])
volatility['CV'] = (volatility['std'] / volatility['mean']) * 100
volatility = volatility.sort_values('CV', ascending=False)
volatility.round(2)


,mean,std,CV
Commodity,,,
Tomato,3724.15,2729.90,73.30
Onion,2214.63,944.79,42.66
Potato,1357.41,519.70,38.29
Rice,3229.15,393.68,12.19
Wheat,2353.47,130.66,5.55


In [4]:
monthly_crop_prices = clean_df.groupby(
    ['Commodity', 'Year', 'Month']
)['Modal_Price'].mean().reset_index()

monthly_crop_prices.columns = ['Commodity', 'Year', 'Month', 'Avg_Modal_Price']
monthly_crop_prices = monthly_crop_prices.sort_values(
    ['Commodity', 'Year', 'Month']
).reset_index(drop=True)

print(monthly_crop_prices.shape)
monthly_crop_prices.head(10)

(67, 4)


,Commodity,Year,Month,Avg_Modal_Price
0,Onion,2023,6,1119.212229
1,Onion,2023,7,1404.686554
2,Onion,2023,8,1691.915848
3,Onion,2023,9,1921.675517
4,Onion,2023,10,2189.736582
5,Onion,2023,11,3543.975133
6,Onion,2023,12,2737.284713
7,Onion,2024,1,1865.163121
8,Onion,2024,2,1559.834785
9,Onion,2024,3,1780.058879


In [5]:
import os
os.makedirs("../data/processed", exist_ok=True)

# File 1 — already built above
monthly_crop_prices.to_csv(
    "../data/processed/monthly_crop_prices.csv", index=False
)

# File 2 — crop summary
crop_summary = clean_df.groupby('Commodity')['Modal_Price'].agg(
    ['mean', 'std', 'min', 'max', 'count']
).round(2).reset_index()
crop_summary['CV'] = (crop_summary['std'] / crop_summary['mean'] * 100).round(2)
crop_summary.columns = [
    'Commodity', 'Mean_Price', 'Std_Price',
    'Min_Price', 'Max_Price', 'Record_Count', 'CV_Percent'
]
crop_summary.to_csv(
    "../data/processed/crop_summary.csv", index=False
)

# File 3 — district price summary
district_summary = clean_df.groupby(
    ['District Name', 'Commodity']
)['Modal_Price'].agg(['mean', 'std', 'count']).round(2).reset_index()
district_summary.columns = [
    'District', 'Commodity', 'Mean_Price', 'Std_Price', 'Record_Count'
]
district_summary = district_summary[
    district_summary['Record_Count'] >= 30
].reset_index(drop=True)
district_summary.to_csv(
    "../data/processed/district_price_summary.csv", index=False
)

print("Files saved successfully")
print("monthly_crop_prices:", monthly_crop_prices.shape)
print("crop_summary:", crop_summary.shape)
print("district_summary:", district_summary.shape)

Files saved successfully
monthly_crop_prices: (67, 4)
crop_summary: (5, 7)
district_summary: (259, 5)


## Day 4 Summary — Aggregation & Time-Series Preparation

### Setup
Loaded clean_mandi_prices.csv (188,080 rows × 10 cols), re-applied datetime 
conversion using format='%Y-%m-%d' (ISO format — pandas writes dates in this 
format when saving to CSV, resolving the M/D/Y ambiguity from Day 3). 
Extracted Year and Month as separate integer columns using `.dt.year` and 
`.dt.month`.

### Part 1 — Datetime understanding
- Confirmed Price Date dtype is datetime64[us] after conversion
- Extracted Year (2023/2024/2025) and Month (1-12) as features
- Understood why feature extraction matters: ML models need numeric columns, 
  not raw datetime objects — Month captures seasonality, Year captures 
  long-term trend

### Part 2 — GroupBy fundamentals
- Learned groupby analogy: sorting receipts into envelopes, then computing 
  per-envelope
- Average Modal_Price by crop: Tomato highest (₹3724), Potato lowest (₹1357)
- Discovered market-level groupby is misleading when mixing crops — 
  Shamli market showed ₹4201 "average" but had only 7 rows, all Onion, 
  making it statistically unreliable
- Correct approach: always groupby ['Market', 'Commodity'] together, 
  never just Market alone for price comparisons
- Many (Market, Commodity) combinations have only 1 row — market-level 
  estimates unreliable for sparse combinations; dashboard should use 
  district-level or show a minimum-data warning

### Part 3 — Monthly aggregation
- Computed monthly average Modal_Price per crop (combining years) and 
  per crop per year separately
- Discovered critical data limitation through investigation:
  - Rice: all data from 2025 only (Apr-Jun) — 3 months, single year
  - Tomato: all data from 2023 only (Jun-Nov) — 6 months, single year
  - Onion/Potato: full coverage across all 3 years (boundary-effect gaps 
    only — e.g., no Jan 2023 because dataset starts Jun 2023)
  - Wheat: 9 months across years
- Conclusion: Rice and Tomato coverage gaps are dataset construction 
  artifacts, NOT real seasonality. Seasonality analysis reliable only 
  for Onion, Potato, Wheat.

### Part 4 — Pivot tables
- Built mean pivot (Commodity × Month → avg price) and count pivot 
  (Commodity × Month → record count)
- Count pivot is a "confidence map" — cells with few records 
  (e.g., Tomato November: 143 records) are less reliable than 
  cells with thousands
- Key seasonal patterns visible:
  - Onion: low in June (₹1192), peaks November (₹3696) — 3x swing
  - Tomato: dramatic July spike (₹6542 vs June's ₹1659) — monsoon effect
  - Wheat: nearly flat (₹2251-₹2509) — MSP price support confirmed

### Part 5 — Volatility analysis
Computed CV (Coefficient of Variation = std/mean × 100) per crop:

| Crop    | Mean (₹) | Std (₹) | CV (%)  |
|---------|----------|---------|---------|
| Tomato  | 3724     | 2730    | 73.30   |
| Onion   | 2215     | 945     | 42.66   |
| Potato  | 1357     | 520     | 38.29   |
| Rice    | 3229     | 394     | 12.19   |
| Wheat   | 2353     | 131     | 5.55    |

- CV preferred over raw std for cross-crop comparison (adjusts for 
  price scale differences)
- Wheat's low CV confirms government MSP intervention
- Tomato's CV based on single year only — treat as indicative, not definitive
- Implications for Day 16 ranking engine: CV will feed directly into 
  "risk/stability" scoring component

### Part 6 — Saved files
- data/processed/monthly_crop_prices.csv (67 rows × 4 cols)
- data/processed/crop_summary.csv (5 rows × 7 cols)
- data/processed/district_price_summary.csv (259 rows × 5 cols, 
  filtered to minimum 30 records per district-crop combination)

### Key concept learned today that wasn't in the original plan
The difference between a dataset coverage gap explained by boundary 
effects (Onion missing Jan 2023 because dataset starts Jun 2023) vs. 
a crop-specific collection artifact (Rice missing all of 2023 and 2024 
despite the dataset covering those years for other crops). This 
distinction matters for honest model validation scope and project 
documentation.